File: 701→800/796. Rotate String/Solution1Liner.cpp
class Solution {
public:
    bool rotateString(string s, string goal) {
        // Check if the lengths of `s` and `goal` are the same.
        // If not, `s` cannot be a rotation of `goal`.
        return s.size() == goal.size() 
            // Concatenate `s` with itself and check if `goal` is a substring of this new string.
            && (s + s).find(goal) != string::npos;
    }
};

/*
Approach:

    This solution checks if `goal` can be obtained by rotating `s`. A string can be rotated to produce all possible 
    rotations by concatenating it with itself. By doing this, all rotations of `s` appear as substrings within `s + s`.
    Therefore, we only need to check if `goal` appears as a substring in `s + s`. 
    Additionally, we confirm that `s` and `goal` have the same length, as different lengths would immediately rule out rotation.

Complexity:

    Time Complexity: O(n), where `n` is the length of `s`. Concatenating `s` with itself and searching for `goal` 
    within this new string both take linear time.

    Space Complexity: O(n) for storing the concatenated string `s + s`.
*/

File: 701→800/796. Rotate String/Solution1Liner.js
/**
 * @param {string} s
 * @param {string} goal
 * @return {boolean}
 */
var rotateString = function (s, goal) {
  // Check if the lengths are equal and if `goal` is a substring of `s + s`
  return s.length === goal.length && (s + s).includes(goal);
};

/* 
Explanation

    The solution first checks if s and goal have the same length, since strings of different lengths can’t be rotations of each other.
    Then, it concatenates s with itself, forming a new string s + s. If goal is a rotation of s, it must appear as a substring in this concatenated string.
        For example, if s = "abcde" and goal = "cdeab", then s + s = "abcdeabcde", and "cdeab" appears as a substring of "abcdeabcde".

Complexity Analysis

    Time Complexity: O(n), where nn is the length of s. The includes method checks for a substring match in O(n) time.
    Space Complexity: O(n) for the temporary concatenated string s + s.
*/
